### ISSUE TITLE
Various keyboard crashes and issues

### ISSUE THREAD

**Issue by poire-z** [koreader/koreader#5632]
* KOReader version: v2019.11-27-g4740ab1

Possibly related to recent refactoring (#5583, #5610), but might be present before.

There's some issue with the Japanese keyboard.

Hitting that yellow key on the emulator causes a crash:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69483833-9935a100-0e2c-11ea-966f-2b7aea9af7ec.png)</kbd>
```
./luajit: frontend/ui/widget/textwidget.lua:135: Wrong text type (expected string)
stack traceback:
        [C]: in function 'assert'
        frontend/ui/widget/textwidget.lua:135: in function 'pre_guard'
        frontend/dbg.lua:41: in function 'updateSize'
        frontend/ui/widget/textwidget.lua:192: in function 'getSize'
        frontend/ui/widget/container/centercontainer.lua:10: in function 'paintTo'
        frontend/ui/widget/container/framecontainer.lua:96: in function 'paintTo'
        frontend/ui/widget/container/inputcontainer.lua:85: in function 'paintTo'
        frontend/ui/widget/horizontalgroup.lua:49: in function 'paintTo'
        frontend/ui/widget/verticalgroup.lua:48: in function 'paintTo'
        frontend/ui/widget/container/centercontainer.lua:24: in function 'paintTo'
        frontend/ui/widget/container/framecontainer.lua:96: in function 'paintTo'
        frontend/ui/widget/container/bottomcontainer.lua:18: in function 'paintTo'
        frontend/ui/widget/container/inputcontainer.lua:85: in function 'paintTo'
        frontend/ui/uimanager.lua:931: in function '_repaint'
```
But on my Kobo, when hitting that same key, I first get a japanese keyboard with all keys blank. Then, when I hit the screen, I get this crash (note the strange reference to the ko_KR keyboard!):
```
./luajit: ./ffi/util.lua:412: attempt to get length of local 'charstring' (a nil value)
stack traceback:
    ./ffi/util.lua:412: in function 'utf8charcode'
    frontend/ui/data/keyboardlayouts/ko_KR_helper.lua:148: in function 'in_consonnant_char'
    frontend/ui/data/keyboardlayouts/ko_KR_helper.lua:270: in function '_should_handle_as_target_char'
    frontend/ui/data/keyboardlayouts/ko_KR_helper.lua:253: in function 'process_char'
    frontend/ui/data/keyboardlayouts/ko_KR_keyboard.lua:41: in function 'addChars'
    frontend/ui/widget/virtualkeyboard.lua:841: in function 'addChar'
    frontend/ui/widget/virtualkeyboard.lua:121: in function 'callback'
    frontend/ui/widget/virtualkeyboard.lua:294: in function 'handleEvent'
    frontend/ui/widget/container/inputcontainer.lua:259: in function 'handleEvent'
    frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
    frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
    ...
    frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
    frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
    frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
    frontend/ui/uimanager.lua:732: in function 'sendEvent'
```
That reference to the korean keyboard might be because I went to the JA keyboard by visiting the KO keyboard - and I remember the KO keyboard wrap the Input object with its specific engine - which might be still there when we go on using another layout.

There's another refresh issue with the taller japanese keyboard: when switching to a smaller keyboard, part of the japanese one does not disappear:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69484005-5b397c80-0e2e-11ea-904f-0ac0b0d557cf.png)</kbd>

Another issue is that the TextEditor plugin does not resize when you switch from a small to a tall keyboard layout - so when EN > JA, the bottom button (Reset, Save, Close) are hidden by the japanese keyboard.

**Comment by NiLuJe** [comment]
I haven't looked at the keyboard since the refactor, but that last one makes sense. IIRC, the keyboard popping up doesn't redraw what's behind it. That worked when there was only a single keyboard, but not for a keyboard swap ;).

**Comment by Frenzie** [comment]
I created those AppImages a bit on a whim almost two years ago, but they're a fantastic way to track regressions. ;-)

tl;dr Looks like there was a suble extra keyboard layout I overlooked in #5610. (Note the 6th column.)

<img src=https://user-images.githubusercontent.com/202757/69484397-1532e780-0e33-11ea-8ed6-6e4f4a3d5d37.png width=45%><img src=https://user-images.githubusercontent.com/202757/69484394-0e0bd980-0e33-11ea-8d7d-077a25b16954.png width=45%>

**Comment by poire-z** [comment]
Working on above issues (refresh on heigt change, text editor resize, korean wrapper unwrap) - except the Japanese crash on tap / layout issues that I'm not familiar with.

**Comment by Frenzie** [comment]
Alright. I should be able to take a look at the keyboard soon, but not tonight.

**Comment by NiLuJe** [comment]
@poire-z: If you want to go the extra mile, both the topmenu & configdialog have a fancy "refresh behind me if I'm smaller than the previous tab" check. But, here, I don't think we need to be that fancy, a simple forced refresh on swap ought to be enough ;).

**Comment by poire-z** [comment]
I won't got the extra mile :) Played with combining prev & new keyboard region, but some other things may move on the screen. So, `UIManager:setDirty("all", "flashui")` it will be, in the rare case one swap from/to the larger japanese keyboard.

**Comment by NiLuJe** [comment]
Yep, sounds good ;). Thanks!
